File: addons/guide-cpp/src/inputs/guide_input_any.cpp
#include "guide_input_any.h"
#include "guide_input_state.h"
#include <godot_cpp/core/math.hpp>

using namespace godot;

void GUIDEInputAny::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_mouse_buttons"), &GUIDEInputAny::get_mouse_buttons);
    ClassDB::bind_method(D_METHOD("set_mouse_buttons", "val"), &GUIDEInputAny::set_mouse_buttons);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "mouse_buttons"), "set_mouse_buttons", "get_mouse_buttons");

    ClassDB::bind_method(D_METHOD("get_mouse_movement"), &GUIDEInputAny::get_mouse_movement);
    ClassDB::bind_method(D_METHOD("set_mouse_movement", "val"), &GUIDEInputAny::set_mouse_movement);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "mouse_movement"), "set_mouse_movement", "get_mouse_movement");

    ClassDB::bind_method(D_METHOD("get_minimum_mouse_movement_distance"), &GUIDEInputAny::get_minimum_mouse_movement_distance);
    ClassDB::bind_method(D_METHOD("set_minimum_mouse_movement_distance", "val"), &GUIDEInputAny::set_minimum_mouse_movement_distance);
    ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "minimum_mouse_movement_distance"), "set_minimum_mouse_movement_distance", "get_minimum_mouse_movement_distance");

    ClassDB::bind_method(D_METHOD("get_joy_buttons"), &GUIDEInputAny::get_joy_buttons);
    ClassDB::bind_method(D_METHOD("set_joy_buttons", "val"), &GUIDEInputAny::set_joy_buttons);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "joy_buttons"), "set_joy_buttons", "get_joy_buttons");

    ClassDB::bind_method(D_METHOD("get_joy_axes"), &GUIDEInputAny::get_joy_axes);
    ClassDB::bind_method(D_METHOD("set_joy_axes", "val"), &GUIDEInputAny::set_joy_axes);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "joy_axes"), "set_joy_axes", "get_joy_axes");

    ClassDB::bind_method(D_METHOD("get_minimum_joy_axis_actuation_strength"), &GUIDEInputAny::get_minimum_joy_axis_actuation_strength);
    ClassDB::bind_method(D_METHOD("set_minimum_joy_axis_actuation_strength", "val"), &GUIDEInputAny::set_minimum_joy_axis_actuation_strength);
    ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "minimum_joy_axis_actuation_strength"), "set_minimum_joy_axis_actuation_strength", "get_minimum_joy_axis_actuation_strength");

    ClassDB::bind_method(D_METHOD("get_keyboard"), &GUIDEInputAny::get_keyboard);
    ClassDB::bind_method(D_METHOD("set_keyboard", "val"), &GUIDEInputAny::set_keyboard);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "keyboard"), "set_keyboard", "get_keyboard");

    ClassDB::bind_method(D_METHOD("get_touch"), &GUIDEInputAny::get_touch);
    ClassDB::bind_method(D_METHOD("set_touch", "val"), &GUIDEInputAny::set_touch);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "touch"), "set_touch", "get_touch");

    ClassDB::bind_method(D_METHOD("_refresh"), &GUIDEInputAny::_refresh);
    ClassDB::bind_method(D_METHOD("_to_string"), &GUIDEInputAny::_to_string);
}

GUIDEInputAny::GUIDEInputAny() {
}

GUIDEInputAny::~GUIDEInputAny() {
}

bool GUIDEInputAny::_needs_reset() const {
    return true;
}

void GUIDEInputAny::_begin_usage() {
    if (_state) {
        if (mouse_movement) _state->connect("mouse_position_changed", Callable(this, "_refresh"));
        if (mouse_buttons) _state->connect("mouse_button_state_changed", Callable(this, "_refresh"));
        if (keyboard) _state->connect("keyboard_state_changed", Callable(this, "_refresh"));
        if (joy_buttons) _state->connect("joy_button_state_changed", Callable(this, "_refresh"));
        if (joy_axes) _state->connect("joy_axis_state_changed", Callable(this, "_refresh"));
        if (touch) _state->connect("touch_state_changed", Callable(this, "_refresh"));
        _refresh();
    }
}

void GUIDEInputAny::_end_usage() {
    if (_state) {
        if (mouse_movement) _state->disconnect("mouse_position_changed", Callable(this, "_refresh"));
        if (mouse_buttons) _state->disconnect("mouse_button_state_changed", Callable(this, "_refresh"));
        if (keyboard) _state->disconnect("keyboard_state_changed", Callable(this, "_refresh"));
        if (joy_buttons) _state->disconnect("joy_button_state_changed", Callable(this, "_refresh"));
        if (joy_axes) _state->disconnect("joy_axis_state_changed", Callable(this, "_refresh"));
        if (touch) _state->disconnect("touch_state_changed", Callable(this, "_refresh"));
    }
}

void GUIDEInputAny::_refresh() {
    if (!_value.is_zero_approx()) return;
    if (!_state) return;

    if (keyboard && _state->is_any_key_pressed()) {
        _value = Vector3(1, 0, 0);
        return;
    }
    if (mouse_buttons && _state->is_any_mouse_button_pressed()) {
        _value = Vector3(1, 0, 0);
        return;
    }
    if (mouse_movement && _state->get_mouse_delta_since_last_frame().length() >= minimum_mouse_movement_distance) {
        _value = Vector3(1, 0, 0);
        return;
    }
    if (joy_buttons && _state->is_any_joy_button_pressed()) {
        _value = Vector3(1, 0, 0);
        return;
    }
    if (joy_axes && _state->is_any_joy_axis_actuated(minimum_joy_axis_actuation_strength)) {
        _value = Vector3(1, 0, 0);
        return;
    }
    if (touch && _state->is_any_finger_down()) {
        _value = Vector3(1, 0, 0);
        return;
    }
    _value = Vector3(0, 0, 0);
}

bool GUIDEInputAny::is_same_as(const Ref<GUIDEInput> &other) const {
    Ref<GUIDEInputAny> o = other;
    if (o.is_null()) return false;
    return o->mouse_buttons == mouse_buttons && 
           o->mouse_movement == mouse_movement &&
           o->joy_buttons == joy_buttons && 
           o->joy_axes == joy_axes &&
           o->keyboard == keyboard && 
           o->touch == touch &&
           Math::abs(o->minimum_mouse_movement_distance - minimum_mouse_movement_distance) < 0.00001 &&
           Math::abs(o->minimum_joy_axis_actuation_strength - minimum_joy_axis_actuation_strength) < 0.00001;
}

String GUIDEInputAny::_editor_name() const {
    return "Any Input";
}

String GUIDEInputAny::_editor_description() const {
    return "Input that triggers if any input from the given device class is given.";
}

GUIDEAction::GUIDEActionValueType GUIDEInputAny::_native_value_type() const {
    return GUIDEAction::GUIDEActionValueType::BOOL;
}

GUIDEInput::DeviceType GUIDEInputAny::_device_type() const {
    int res = NONE;
    if (joy_axes || joy_buttons) res |= JOY;
    if (mouse_buttons || mouse_movement) res |= MOUSE;
    if (keyboard) res |= KEYBOARD;
    if (touch) res |= TOUCH;
    return (DeviceType)res;
}

File: addons/GuideCS/Inputs/GUIDEInputAny.cs
using System;
using System.Collections.Generic;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEInputAny : GUIDEInput {
	// Deprecated properties
	[Obsolete("Deprecated, use mouse_buttons instead")]
	public bool Mouse { get => MouseButtons; set => MouseButtons = value; }
	[Obsolete("Deprecated, use joy_buttons instead")]
	public bool Joy { get => JoyButtons; set => JoyButtons = value; }

	[Export] public bool MouseButtons { get; set; } = false;
	[Export] public bool MouseMovement { get; set; } = false;
	[Export] public float MinimumMouseMovementDistance { get; set; } = 1.0f;
	[Export] public bool JoyButtons { get; set; } = false;
	[Export] public bool JoyAxes { get; set; } = false;
	[Export] public float MinimumJoyAxisActuationStrength { get; set; } = 0.2f;
	[Export] public bool Keyboard { get; set; } = false;
	[Export] public bool Touch { get; set; } = false;

	public override bool _NeedsReset() => true;

	public override void _BeginUsage() {
		if (MouseMovement) _State.MousePositionChanged += _Refresh;
		if (MouseButtons) _State.MouseButtonStateChanged += _Refresh;
		if (Keyboard) _State.KeyboardStateChanged += _Refresh;
		if (JoyButtons) _State.JoyButtonStateChanged += _Refresh;
		if (JoyAxes) _State.JoyAxisStateChanged += _Refresh;
		if (Touch) _State.TouchStateChanged += _Refresh;

		_Refresh();
	}

	public override void _EndUsage() {
		if (MouseMovement) _State.MousePositionChanged -= _Refresh;
		if (MouseButtons) _State.MouseButtonStateChanged -= _Refresh;
		if (Keyboard) _State.KeyboardStateChanged -= _Refresh;
		if (JoyButtons) _State.JoyButtonStateChanged -= _Refresh;
		if (JoyAxes) _State.JoyAxisStateChanged -= _Refresh;
		if (Touch) _State.TouchStateChanged -= _Refresh;
	}

	private void _Refresh() {
		if (!_Value.IsZeroApprox()) return;

		if (Keyboard && _State.IsAnyKeyPressed()) {
			_Value = Vector3.Right;
			return;
		}

		if (MouseButtons && _State.IsAnyMouseButtonPressed()) {
			_Value = Vector3.Right;
			return;
		}

		if (MouseMovement && _State.GetMouseDeltaSinceLastFrame().Length() >= MinimumMouseMovementDistance) {
			_Value = Vector3.Right;
			return;
		}

		if (JoyButtons && _State.IsAnyJoyButtonPressed()) {
			_Value = Vector3.Right;
			return;
		}

		if (JoyAxes && _State.IsAnyJoyAxisActuated(MinimumJoyAxisActuationStrength)) {
			_Value = Vector3.Right;
			return;
		}

		if (Touch && _State.IsAnyFingerDown()) {
			_Value = Vector3.Right;
			return;
		}

		_Value = Vector3.Zero;
	}

	public override bool IsSameAs(GUIDEInput other) {
		return other is GUIDEInputAny any &&
			   MouseButtons == any.MouseButtons &&
			   MouseMovement == any.MouseMovement &&
			   JoyButtons == any.JoyButtons &&
			   JoyAxes == any.JoyAxes &&
			   Keyboard == any.Keyboard &&
			   Touch == any.Touch &&
			   Mathf.IsEqualApprox(MinimumMouseMovementDistance, any.MinimumMouseMovementDistance) &&
			   Mathf.IsEqualApprox(MinimumJoyAxisActuationStrength, any.MinimumJoyAxisActuationStrength);
	}

	public override string _EditorName() => "Any Input";

	public override string _EditorDescription() => "Input that triggers if any input from the given device class is given.";

	public override GUIDEAction.GUIDEActionValueType _NativeValueType() => GUIDEAction.GUIDEActionValueType.Bool;

	public override Godot.Collections.Array<Godot.Collections.Dictionary> _GetPropertyList() {
		return new Godot.Collections.Array<Godot.Collections.Dictionary> {
			new Godot.Collections.Dictionary
			{
				{ "name", "Mouse" },
				{ "type", (int)Variant.Type.Bool },
				{ "usage", (int)PropertyUsageFlags.NoEditor }
			},
			new Godot.Collections.Dictionary {
			   { "name", "Joy" },
			   { "type", (int)Variant.Type.Bool },
			   { "usage", (int)PropertyUsageFlags.NoEditor }
		   }
		};
	}

	public override DeviceType _DeviceType() {
		DeviceType result = DeviceType.None;
		if (JoyAxes || JoyButtons) result |= DeviceType.Joy;
		if (MouseButtons || MouseMovement) result |= DeviceType.Mouse;
		if (Keyboard) result |= DeviceType.Keyboard;
		if (Touch) result |= DeviceType.Touch;
		return result;
	}
}
